eventuals/expected.h

```diff
@@ -25,30 +25,66 @@ Overloaded(Fs...) -> Overloaded<Fs...>;
 
 ////////////////////////////////////////////////////////////////////////
 
-template <typename Value_, typename = std::exception_ptr>
-class Expected;
+// Tag type to identify the result of 'Unexpected(...)'.
+struct _Unexpected {
+  _Unexpected() = delete;
+};
+
+////////////////////////////////////////////////////////////////////////
 
 template <typename Value_>
-class Expected<Value_, std::exception_ptr> {
+class _Expected {
  public:
+  // Providing 'ValueFrom' in order to compose with other eventuals.
+  template <typename Arg>
+  using ValueFrom = Value_;
+
   // NOTE: providing a default constructor so this type can be used as
   // a type parameter to 'std::promise' which fails on MSVC (see
   // https://stackoverflow.com/questions/28991694 which also ran into
   // this issue; very surprised this hasn't been fixed, apparently
   // there aren't many users of 'std::promise' on Windows).
-  Expected() {}
-
-  template <typename T>
-  Expected(T&& t)
+  _Expected() {}
+
+  template <
+      typename T,
+      std::enable_if_t<
+          std::is_convertible_v<T, Value_>,
+          int> = 0>
+  _Expected(T&& t)
     : variant_(std::forward<T>(t)) {}
 
-  Expected(Expected&& that) = default;
-  Expected(Expected& that) = default;
-  Expected(const Expected& that) = default;
-
-  Expected& operator=(Expected&& that) = default;
-  Expected& operator=(Expected& that) = default;
-  Expected& operator=(const Expected& that) = default;
+  _Expected(std::exception_ptr exception)
+    : variant_(std::move(exception)) {}
+
+  _Expected(_Expected&& that) = default;
+  _Expected(_Expected& that) = default;
+  _Expected(const _Expected& that) = default;
+
+  _Expected& operator=(_Expected&& that) = default;
+  _Expected& operator=(_Expected& that) = default;
+  _Expected& operator=(const _Expected& that) = default;
+
+  template <typename U>
+  _Expected(_Expected<U>&& that)
+    : variant_(std::move(that).template convert<Value_>()) {}
+
+  template <typename U>
+  _Expected(const _Expected<U>& that)
+    : variant_(that.template convert<Value_>()) {}
+
+  // Providing 'k()' in order to compose with other eventuals.
+  template <typename Arg, typename K>
+  auto k(K k) && {
+    return Eventual<Value_>([variant = std::move(variant_)](auto& k) mutable {
```

Comment:
should it propagate Fail if we start an Eventual from Stop for example?

Reply:
This won't propagate a `Fail()` if it gets a `Stop()`, it'll just propagate the `Stop()`. It'll only propagate `Fail()` if it gets a `Start()`.

Reply:
I understand, I'm just speaking about what if we do in Stop k.Fail(std::move(std::get<1>(variant))) in case when variant.index() != 0 if it exists

Reply:
As stated, 'Stop()' will continue to get propagated, regardless of what `variant.index()` is. If we propagate `Fail()` instead then someone downstream might attempt to recover from the error instead of knowing that we want to actually stop, hence we propagate `Stop()`. I'll add this as a comment.